File: OJ practices/LeetCode题目集/2. Add Two Numbers(medium).java
/*
 * @Autor: SourDumplings
 * @Date: 2019-12-08 22:45:03
 * @Link: https://github.com/SourDumplings/
 * @Email: changzheng300@foxmail.com
 * @Description: https://leetcode.com/problems/add-two-numbers/
 */

/**
* Definition for singly-linked list.
* public class ListNode {
*     int val;
*     ListNode next;
*     ListNode(int x) { val = x; }
* }
*/
class Solution
{
    public ListNode addTwoNumbers(ListNode l1, ListNode l2)
    {
        ListNode head = new ListNode(0);
        head.next = null;
        ListNode p1 = l1, p2 = l2, p = head;
        boolean flag = false;
        int v1, v2;
        while (p1 != null || p2 != null)
        {
            p.next = new ListNode(0);
            p = p.next;
            p.next = null;
            if (p1 != null)
            {
                v1 = p1.val;
                p1 = p1.next;
            }
            else
            {
                v1 = 0;
            }
            if (p2 != null)
            {
                v2 = p2.val;
                p2 = p2.next;
            }
            else
            {
                v2 = 0;
            }
            p.val = v1 + v2;
            if (flag)
            {
                ++p.val;
                flag = false;
            }
            if (p.val >= 10)
            {
                p.val -= 10;
                flag = true;
            }
        }
        if (flag)
        {
            p.next = new ListNode(1);
            p.next.next = null;
        }
        return head.next;
    }
}

File: OJ practices/LeetCode题目集/2. Add Two Numbers(medium).cpp
/**
 * @file 2. Add Two Numbers(medium).cpp
 * @author SourDumplings (sourdumplings@qq.com)
 * @brief https://leetcode.cn/problems/add-two-numbers/?envType=study-plan-v2&envId=top-interview-150
 * @version 1.0.0
 * @date 2024-11-14
 *
 * @copyright Copyright (c) 2024 SourDumplings
 *
 */

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
	ListNode* addTwoNumbers(ListNode* l1, ListNode* l2)
	{
		bool isCarryOver = false;
		ListNode* p1	 = l1;
		ListNode* p2	 = l2;
		ListNode* h		 = new ListNode(0);
		ListNode* p		 = h;
		while (p1 && p2)
		{
			p->next		= new ListNode(0);
			p			= p->next;
			p->val		= p1->val + p2->val + (isCarryOver ? 1 : 0);
			isCarryOver = false;
			if (10 <= p->val)
			{
				isCarryOver	 = true;
				p->val		%= 10;
			}
			p1 = p1->next;
			p2 = p2->next;
		}
		while (p1)
		{
			p->next		= new ListNode(0);
			p			= p->next;
			p->val		= p1->val + (isCarryOver ? 1 : 0);
			isCarryOver = false;
			if (10 <= p->val)
			{
				isCarryOver	 = true;
				p->val		%= 10;
			}
			p1 = p1->next;
		}
		while (p2)
		{
			p->next		= new ListNode(0);
			p			= p->next;
			p->val		= p2->val + (isCarryOver ? 1 : 0);
			isCarryOver = false;
			if (10 <= p->val)
			{
				isCarryOver	 = true;
				p->val		%= 10;
			}
			p2 = p2->next;
		}
		if (isCarryOver) { p->next = new ListNode(1); }

		p = h;
		h = p->next;
		delete p;
		return h;
	}
};
